File: Accounting/src/main/java/com/experiment/accounting/service/ConstantService.java
package com.experiment.accounting.service;

import com.experiment.accounting.domain.base.BaseConstantModel;
import com.experiment.accounting.entity.base.BaseConstantEntity;
import com.experiment.accounting.exception.BusinessException;
import com.experiment.accounting.exception.BusinessExceptions;
import com.experiment.accounting.mapper.ConstantMapper;
import com.experiment.accounting.repository.base.EntityRepository;
import com.experiment.accounting.service.base.EntityService;
import com.experiment.accounting.utils.ConstantUtils;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.cache.CacheManager;
import org.springframework.cache.annotation.Cacheable;
import org.springframework.data.domain.Example;
import org.springframework.stereotype.Service;

import java.util.List;
import java.util.Optional;
import java.util.stream.Collectors;

@Service
public class ConstantService extends EntityService {

    private final ConstantMapper constantMapper;
    private final CacheManager cacheManager;

    public ConstantService(ConstantMapper constantMapper, CacheManager cacheManager, EntityRepository<BaseConstantEntity> entityRepository) {
        super(entityRepository);
        this.cacheManager = cacheManager;
        this.constantMapper = constantMapper;
    }

    @Cacheable(value = "findReference", key = "#name + #c.getName()")
    public <T extends BaseConstantEntity> Optional<T> find(String name, Class<T> c) {
        T instance = createInstance(c);
        instance.setName(name);
        return entityRepository.findOne(Example.of(instance));
    }

    @Cacheable(value = "findReferenceByName", key = "#constantName")
    public <T extends BaseConstantModel> List<T> findAllByConstantName(String constantName) {
        return findAll(ConstantUtils.getReferenceByName(constantName).getClazz())
                .stream()
                .map(constantEntity -> (T) constantMapper.toModel(constantEntity))
                .collect(Collectors.toList());
    }

    public void deleteByConstantNameAndId(String constantName, Long id) {
        delete(id, ConstantUtils.getReferenceByName(constantName).getClazz());
        cacheManager.getCache("findReferenceByName").evictIfPresent(constantName);
    }

    public <T extends BaseConstantEntity> T saveByConstantName(String constantName, BaseConstantModel constantModel) {
        if (constantModel.getId() != null)
            throw new BusinessException("If id field is not null, try to update " + constantName);
        T t = save(constantMapper.toEntity(constantModel, ConstantUtils.getReferenceByName(constantName).getClazz()));
        cacheManager.getCache("findReferenceByName").evictIfPresent(constantName);
        return t;
    }

    public <T extends BaseConstantEntity> T updateByConstantName(String constantName, BaseConstantModel constantModel) {
        if (constantModel.getId() == null)
            throw new BusinessException("If id field is null, try to create " + constantName);
        T t = save(constantMapper.toEntity(constantModel, ConstantUtils.getReferenceByName(constantName).getClazz()));
        cacheManager.getCache("findReferenceByName").evictIfPresent(constantName);
        return t;

    }

}

File: Inventory/src/main/kotlin/com/experiment/inventory/service/ConstantService.kt
package com.experiment.inventory.service

import com.experiment.inventory.domain.base.BaseConstantModel
import com.experiment.inventory.entity.base.BaseConstantEntity
import com.experiment.inventory.exception.BusinessException
import com.experiment.inventory.mapper.ConstantMapper
import com.experiment.inventory.repository.EntityRepository
import com.experiment.inventory.service.base.EntityService
import com.experiment.inventory.utils.ConstantUtils
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.cache.CacheManager
import org.springframework.cache.annotation.Cacheable
import org.springframework.data.domain.Example
import org.springframework.stereotype.Service
import java.util.*
import kotlin.reflect.KClass

@Service
class ConstantService<T>
@Autowired constructor(
    var entityRepository:EntityRepository<T>,
    var cacheManager: CacheManager,
    var constantMapper: ConstantMapper): EntityService<T>() where T:BaseConstantEntity {


    @Cacheable(value = ["findReference"], key = "#name + #c.getName()")
    fun find(name: String, c: KClass<T>): Optional<T> {
        val instance = createInstance(c)
        instance.name = name
        return entityRepository.findOne(Example.of(instance))
    }

    @Cacheable(value = ["findReferenceByName"], key = "#constantName")
    fun  findAllByConstantName(constantName: String): List<BaseConstantModel> {
        return findAll(ConstantUtils.getReferenceByName(constantName).clazz as KClass<T>)
            .map{ constantEntity -> constantMapper.toModel(constantEntity)}.toList()
    }

    fun deleteByConstantNameAndId(constantName: String, id: Long) {
        delete(id, ConstantUtils.getReferenceByName(constantName).clazz as KClass<T>)
        cacheManager.getCache("findReferenceByName")?.evictIfPresent(constantName)
    }

    fun saveByConstantName(constantName: String, constantModel: BaseConstantModel): T? {
        if (constantModel.id != null) throw BusinessException("ID is not null, please try to update constant ($constantName)")
        val entity: T? = save(constantMapper.toEntity(constantModel, ConstantUtils.getReferenceByName(constantName).clazz as KClass<T>))
        cacheManager.getCache("findReferenceByName")?.evictIfPresent(constantName)
        return entity
    }

    fun updateByConstantName(constantName: String, constantModel: BaseConstantModel): T? {
        if (constantModel.id == null) throw BusinessException("ID is null, try to create constant ($constantName)")
        val reference = ConstantUtils.getReferenceByName(constantName);
        val clazz = reference.clazz as KClass<T>
        val entity: T? = save(constantMapper.toEntity(constantModel, clazz))
        cacheManager.getCache("findReferenceByName")?.evictIfPresent(constantName)
        return entity
    }
}